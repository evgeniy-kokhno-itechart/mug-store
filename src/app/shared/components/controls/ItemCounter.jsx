import React, { Component } from 'react';
import { FontAwesomeIcon } from '@fortawesome/react-fontawesome';
import { PropTypes } from 'prop-types';

class ItemCounter extends Component {
  handleDecrementClick = () => {
    const { itemId, onDecrementClick } = this.props;
    onDecrementClick(itemId, 1);
  };

  handleIncrementClick = () => {
    const { itemId, onIncrementClick } = this.props;
    onIncrementClick(itemId, 1);
  };

  handleCountChange = (e) => {
    const { itemId, onCountChange } = this.props;
    const val = e.currentTarget.value ? parseInt(e.currentTarget.value, 10) : 1;
    // ternary operator was added to avoid NaN when input is empty
    onCountChange(itemId, val);
  };

  render() {
    const { count } = this.props;
    return (
      <div className="input-group input-group-sm mx-auto flex-nowrap item-counter">
        <button
          type="button"
          className={`item-counter__btn input-group-prepend btn btn-secondary btn-sm p-1 p-lg-2 ${count > 1 ? '' : 'disabled'}`}
          onClick={this.handleDecrementClick}
        >
          <FontAwesomeIcon icon="minus" className="my-auto" />
        </button>

        <input className="form-control inline text-center p-1 p-lg-2" type="text" value={count} onChange={this.handleCountChange} />

        <button
          type="button"
          className="item-counter__btn input-group-append btn btn-secondary btn-sm p-1 p-lg-2"
          onClick={this.handleIncrementClick}
        >
          <FontAwesomeIcon icon="plus" className="my-auto" />
        </button>
      </div>
    );
  }
}

ItemCounter.propTypes = {
  count: PropTypes.number.isRequired,
  itemId: PropTypes.string.isRequired,
  onIncrementClick: PropTypes.func.isRequired,
  onDecrementClick: PropTypes.func.isRequired,
  onCountChange: PropTypes.func.isRequired,
};

export default ItemCounter;
