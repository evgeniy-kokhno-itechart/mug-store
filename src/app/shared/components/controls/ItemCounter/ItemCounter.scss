@import "Variables";

.item-counter {
  display: flex;
  max-width: 7rem;

  &__input {
    font: inherit;
    max-width: 35%;
    text-align: center;
  }

  &__btn {
    line-height: 1.2rem;
    font-size: 0.75rem;
    min-width: 1.7rem;
    padding: $padding-vertical-small;
    color: $color-white;
    border: 1px solid $color-gray-light;
    background-image: linear-gradient(to right bottom, rgba($color-gray-dark, 0.8), rgba($color-gray-dark, 1));
    
    &:disabled {
      background-image: linear-gradient(to right bottom, rgba($color-gray-light-2, 0.8), rgba($color-gray-light-2, 1));
    }
    
  }
}