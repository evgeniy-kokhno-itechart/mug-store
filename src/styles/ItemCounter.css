#bootstrap-overrides .item_counter__btn {
  line-height: 1.2em;
  font-size: 0.75rem;
}
