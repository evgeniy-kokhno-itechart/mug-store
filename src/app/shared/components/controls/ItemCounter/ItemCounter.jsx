import React, { Component } from 'react';
import { FontAwesomeIcon } from '@fortawesome/react-fontawesome';
import { PropTypes } from 'prop-types';
import './ItemCounter.scss';

class ItemCounter extends Component {
  handleInputEdited = (e) => {
    // ternary operator was added to avoid NaN when input is empty
    const newValue = e.target.value;
    const newQuantity = (newValue && newValue > 0) ? parseInt(newValue, 10) : 1;
    this.handleQuantityChange(newQuantity);
  }

  handleDecremented = () => {
    let newQuantity = this.props.item.quantity;
    newQuantity--;
    this.handleQuantityChange(newQuantity);
  };

  handleIncremented = () => {
    let newQuantity = this.props.item.quantity;
    newQuantity++;
    this.handleQuantityChange(newQuantity);
  };

  handleQuantityChange = (newQuantity) => {
    this.props.onQuantityChanged({ ...this.props.item, quantity: newQuantity });
  };

  render() {
    const { item } = this.props;
    return (
      <div className="item-counter">
        <button
          type="button"
          className='item-counter__btn counter-btn-decrease focusable'
          onClick={this.handleDecremented}
          disabled={item.quantity <= 1 ? 'disabled' : ''}
        >
          <FontAwesomeIcon icon="minus" />
        </button>

        <input className="item-counter__input focusable" type="text" value={item.quantity} onChange={this.handleInputEdited} />

        <button
          type="button"
          className="item-counter__btn counter-btn-increase focusable"
          onClick={this.handleIncremented}
        >
          <FontAwesomeIcon icon="plus" />
        </button>
      </div>
    );
  }
}

ItemCounter.propTypes = {
  item: PropTypes.PropTypes.shape({
    id: PropTypes.string,
    quantity: PropTypes.number,
  }).isRequired,
  onQuantityChanged: PropTypes.func.isRequired,
};

export default ItemCounter;
