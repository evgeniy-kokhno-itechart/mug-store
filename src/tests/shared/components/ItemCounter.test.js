/* eslint-disable no-undef */
import React from 'react';

import { configure, shallow } from 'enzyme';
import Adapter from 'enzyme-adapter-react-16';

import ItemCounter from '../../../app/shared/components/controls/ItemCounter';

configure({ adapter: new Adapter() });

describe('<ItemCounter />', () => {
  let ItemCounterWrapper;
  let onQuantityChanged;
  const fakeItem = { id: '1', quantity: 1 };

  beforeEach(() => {
    onQuantityChanged = jest.fn();
    ItemCounterWrapper = shallow(
      <ItemCounter
        item={fakeItem}
        onQuantityChanged={onQuantityChanged}
      />,
    );
  });

  it('renders with decrement button disabled if count=1', () => {
    expect(ItemCounterWrapper).toMatchSnapshot();
  });

  it('renders with decrement button enabled if count is 2 (i.e. greater than 1)', () => {
    ItemCounterWrapper.setProps({ item: { id: '1', quantity: 2 } });
    expect(ItemCounterWrapper).toMatchSnapshot();
  });

  it('calls onQuantityChanged with increased by 1 value on increment button click', () => {
    ItemCounterWrapper.find('.input-group-append').simulate('click');
    expect(onQuantityChanged).toHaveBeenCalledWith({ id: '1', quantity: 2 });
  });

  it('calls onQuantityChanged with decreased by 1 value on decrement button click', () => {
    ItemCounterWrapper.setProps({ item: { id: '1', quantity: 3 } });
    ItemCounterWrapper.find('.input-group-prepend').simulate('click');
    expect(onQuantityChanged).toHaveBeenCalledWith({ id: '1', quantity: 2 });
  });

  it('calls onQuantityChanged with 15 on input changed with 15', () => {
    ItemCounterWrapper.find('input').simulate('change', { target: { name: '', value: '15' } });
    expect(onQuantityChanged).toHaveBeenCalledWith({ id: '1', quantity: 15 });
  });
});
