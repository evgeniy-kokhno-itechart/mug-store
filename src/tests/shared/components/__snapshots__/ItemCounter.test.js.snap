// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<ItemCounter /> renders with decrement button disabled if count=1 1`] = `
<div
  className="item-counter"
>
  <button
    className="item-counter__btn counter-btn-decrease focusable"
    disabled="disabled"
    onClick={[Function]}
    type="button"
  >
    <FontAwesomeIcon
      border={false}
      className=""
      fixedWidth={false}
      flip={null}
      icon="minus"
      inverse={false}
      listItem={false}
      mask={null}
      pull={null}
      pulse={false}
      rotation={null}
      size={null}
      spin={false}
      symbol={false}
      title=""
      transform={null}
    />
  </button>
  <input
    className="item-counter__input focusable"
    onChange={[Function]}
    type="text"
    value={1}
  />
  <button
    className="item-counter__btn counter-btn-increase focusable"
    onClick={[Function]}
    type="button"
  >
    <FontAwesomeIcon
      border={false}
      className=""
      fixedWidth={false}
      flip={null}
      icon="plus"
      inverse={false}
      listItem={false}
      mask={null}
      pull={null}
      pulse={false}
      rotation={null}
      size={null}
      spin={false}
      symbol={false}
      title=""
      transform={null}
    />
  </button>
</div>
`;

exports[`<ItemCounter /> renders with decrement button enabled if count is 2 (i.e. greater than 1) 1`] = `
<div
  className="item-counter"
>
  <button
    className="item-counter__btn counter-btn-decrease focusable"
    disabled=""
    onClick={[Function]}
    type="button"
  >
    <FontAwesomeIcon
      border={false}
      className=""
      fixedWidth={false}
      flip={null}
      icon="minus"
      inverse={false}
      listItem={false}
      mask={null}
      pull={null}
      pulse={false}
      rotation={null}
      size={null}
      spin={false}
      symbol={false}
      title=""
      transform={null}
    />
  </button>
  <input
    className="item-counter__input focusable"
    onChange={[Function]}
    type="text"
    value={2}
  />
  <button
    className="item-counter__btn counter-btn-increase focusable"
    onClick={[Function]}
    type="button"
  >
    <FontAwesomeIcon
      border={false}
      className=""
      fixedWidth={false}
      flip={null}
      icon="plus"
      inverse={false}
      listItem={false}
      mask={null}
      pull={null}
      pulse={false}
      rotation={null}
      size={null}
      spin={false}
      symbol={false}
      title=""
      transform={null}
    />
  </button>
</div>
`;
